stan/math/opencl/kernel_generator/block.hpp

```diff
@@ -0,0 +1,225 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_BLOCK_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_BLOCK_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/scal/err/throw_domain_error.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation_cl_lhs.hpp>
+#include <stan/math/opencl/kernel_generator/as_operation_cl.hpp>
+#include <stan/math/opencl/kernel_generator/is_valid_expression.hpp>
+#include <string>
+#include <type_traits>
+#include <set>
+#include <utility>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents submatrix block in kernel generator expressions.
+ * @tparam Derived derived type
+ * @tparam T type of argument
+ */
+template <typename T>
+class block_
+    : public operation_cl_lhs<block_<T>,
+                              typename std::remove_reference_t<T>::Scalar, T> {
+ public:
+  using Scalar = typename std::remove_reference_t<T>::Scalar;
+  using base = operation_cl_lhs<block_<T>, Scalar, T>;
+  using base::var_name;
+
+ protected:
+  int start_row_, start_col_, rows_, cols_;
+  using base::arguments_;
+
+ public:
+  /**
+   * Constructor
+   * @param a expression
+   * @param start_row first row of block
+   * @param start_col first column of a block
+   * @param rows number of rows in block
+   * @param cols number of columns in block
+   */
+  block_(T&& a, int start_row, int start_col, int rows, int cols)
+      : base(std::forward<T>(a)),
+        start_row_(start_row),
+        start_col_(start_col),
+        rows_(rows),
+        cols_(cols) {
+    if ((a.rows() != base::dynamic && (start_row + rows) > a.rows())
+        || (a.cols() != base::dynamic && (start_col + cols) > a.cols())) {
+      throw_domain_error("block", "block of \"a\"", " is out of bounds", "");
+    }
+  }
+
+  /**
+   * generates kernel code for this and nested expressions.
+   * @param[in,out] generated set of already generated operation_cls
+   * @param ng name generator for this kernel
+   * @param i row index variable name
```

Comment:
These params dont match the actual function signature. Also please start the comment with a capital letter.